examples/seismic/tti/wavesolver.py

```diff
@@ -27,33 +27,51 @@ def __init__(self, model, source, receiver,
         self.receiver = receiver
 
         self.time_order = time_order
-        self.space_order = space_order/2
-
-        # Time step can be \sqrt{3}=1.73 bigger with 4th order
+        self.space_order = space_order
         self.dt = self.model.critical_dt
-        if self.time_order == 4:
-            self.dt *= 1.73
 
         # Cache compiler options
         self._kwargs = kwargs
 
     @cached_property
-    def op_fwd(self):
+    def op_fwd_shifted(self):
```

Comment:
So instead of having 4 *properties* ``op_fwd_shifted, op_fwd_shifted_save, op_fwd_centered, op_fwd_centered_save``, can we not have one *method* taking two optional parameters ``kernel`` and ``save`` ?

Reply:
No idea how to make it a cached property that way

Reply:
10K sources same model, no need to regenerate the operator

Reply:
OK, I see.

What you're looking for is called "memoization" (not a Python-specific concept).
Here a Python example:
https://stackoverflow.com/questions/1988804/what-is-memoization-and-how-can-i-use-it-in-python

It's much much simpler than it seems/sounds

I recommend to adopt the decorator solution (e.g., @memoized) and put it into ``devito/utils.py`` . If too complicated, we can look at this together early next week

Reply:
> and put it into devito/utils.py
Then it's not an example anymore, couldn't disagree more this is gonna  make `examples` a nightmare. Remember that most user will NOT be computer scientist and have about 5-10 min to look at Devito, if I need more than that knowing the software we are shooting ourselves in the leg.
Examples are showcasing what we can do with devito and some repetition is not a problem as you see what's waht at first glance without having to dig into stuff.

Reply:
> Then it's not an example anymore

The subject was the ``memoizer`` ... and ``utils`` was definitely the right place for it. A ``memoizer`` is a general thing.

So... Are you really saying that putting a @memoize decorator on top of a function would make an example less readable?

@mlange05, can you take a look at this?

Reply:
:+1: for `@memoize` - that's exactly what the solver object is effectively doing. It should be a copy-and-past job to provide the decorator (properly linked of course).

Reply:
> It should be a copy-and-past job to provide the decorator

So repalce a copy paste by a copy paste.....